sql/inspect: filter SHOW INSPECT ERRORS by job payload

spilchen opened the pull request:
Previously, `SHOW INSPECT ERRORS FOR TABLE` would select the most recent completed INSPECT job that had errors for that table by joining with the `inspect_errors` table. This approach had a flaw: if a more recent job inspected the table but found no errors, it would be ignored in favor of an older job with errors.

This changes that by query the INSPECT job payload to determine which jobs actually inspected a given table.

Informs #148287

Release note: none
Epic: CRDB-30356

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/154573)
<!-- Reviewable:end -->

spilchen committed:
sql/inspect: filter SHOW INSPECT ERRORS by job payload

Previously, SHOW INSPECT ERRORS FOR TABLE <table> would select the most
recent completed INSPECT job that had errors for that table by joining
with the inspect_errors table. This approach had a flaw: if a more
recent job inspected the table but found no errors, it would be ignored
in favor of an older job with errors.

This changes that by query the INSPECT job payload to determine which
jobs actually inspected a given table.

Informs #148287

Release note: none
Epic: CRDB-30356

rafiss reviewed (approved):
lgtm! could you get a review from @bghal as well since he last worked on this?

spilchen commented on pkg/sql/delegate/show_inspect_errors.go:
For performance, I don't think it matters. The `job_id` is a trivial CTE. So, I'm guessing the optimizer can easily optimize its usage. So, it probably ends up equivalent to a join. I would argue this is easier to understand than a JOIN, but that's just a personal preference. I am fine with changing this if you feel strongly. Let me know either way.
